File: lab/python/arrays_e_listas/inversao.py
def inverter1(lista):
    if not lista:
        return None
    return list(reversed(lista))


def inverter2(lista):
    if not lista:
        return None
    tamanho = len(lista)

    reverso = []

    for i in range(tamanho-1, -1, -1):
        reverso.append(lista[i])

    return reverso


def inverter3(lista):
    if not lista:
        return None

    tamanho = len(lista)

    for i in range(tamanho // 2):
        lista[i], lista[tamanho - 1 - i] = lista[tamanho - 1 - i], lista[i]

    return lista


def inverter4(lista):
    if not lista:
        return None
    return lista[::-1]


lista = [1, 2, 3, 4]
print(inverter1(lista.copy()))
print(inverter2(lista.copy()))
print(inverter3(lista.copy()))
print(inverter4(lista.copy()))

File: lab/javascript/arrays_e_listas/inversao.js
function inverter1(lista) {
  if (!lista || lista.length === 0) return null;
  return [...lista].reverse(); // copia e inverte
}

function inverter2(lista) {
  if (!lista || lista.length === 0) return null;
  const reverso = [];
  for (let i = lista.length - 1; i >= 0; i--) {
    reverso.push(lista[i]);
  }
  return reverso;
}

function inverter3(lista) {
  if (!lista || lista.length === 0) return null;
  const arr = [...lista]; // cria cópia para não modificar original
  const tamanho = arr.length;

  for (let i = 0; i < Math.floor(tamanho / 2); i++) {
    const temp = arr[i];
    arr[i] = arr[tamanho - 1 - i];
    arr[tamanho - 1 - i] = temp;
  }

  return arr;
}

function inverter4(lista) {
  if (!lista || lista.length === 0) return null;
  return lista.slice().reverse(); // mesma ideia do slicing Python [::-1]
}

// -------------------- Teste --------------------
const lista = [1, 2, 3, 4];
console.log(inverter1(lista));
console.log(inverter2(lista));
console.log(inverter3(lista));
console.log(inverter4(lista));
